File: 3.x.x/crates-utopia/query/src/compiler.rs
//! PHP `Utopia\Query\Compiler`.

use crate::error::QueryError;
use crate::query::Query;

/// PHP `Utopia\Query\Compiler`.
pub trait Compiler {
    fn compile_filter(&mut self, query: &Query) -> Result<String, QueryError>;
    fn compile_order(&mut self, query: &Query) -> Result<String, QueryError>;
    fn compile_limit(&mut self, query: &Query) -> Result<String, QueryError>;
    fn compile_offset(&mut self, query: &Query) -> Result<String, QueryError>;
    fn compile_select(&mut self, query: &Query) -> Result<String, QueryError>;
    fn compile_cursor(&mut self, query: &Query) -> Result<String, QueryError>;
    fn compile_aggregate(&mut self, query: &Query) -> Result<String, QueryError>;
    fn compile_group_by(&mut self, query: &Query) -> Result<String, QueryError>;
    fn compile_join(&mut self, query: &Query) -> Result<String, QueryError>;
}

File: packages/query/src/Compiler.php
<?php

namespace Utopia\Query;

interface Compiler
{
    /**
     * Compile a filter query (equal, greaterThan, contains, between, spatial, vector, logical, etc.)
     */
    public function compileFilter(Query $query): string;

    /**
     * Compile an order query (orderAsc, orderDesc, orderRandom)
     */
    public function compileOrder(Query $query): string;

    /**
     * Compile a limit query
     */
    public function compileLimit(Query $query): string;

    /**
     * Compile an offset query
     */
    public function compileOffset(Query $query): string;

    /**
     * Compile a select query
     */
    public function compileSelect(Query $query): string;

    /**
     * Compile a cursor query (cursorAfter, cursorBefore)
     */
    public function compileCursor(Query $query): string;

    /**
     * Compile an aggregate query (count, sum, avg, min, max)
     */
    public function compileAggregate(Query $query): string;

    /**
     * Compile a group by query
     */
    public function compileGroupBy(Query $query): string;

    /**
     * Compile a join query (join, leftJoin, rightJoin, crossJoin)
     */
    public function compileJoin(Query $query): string;
}
